compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -10609,4 +10601,130 @@ protected void addTransactionInternalModuleImport() {
             env.enclPkg.symbol.imports.add(importDcl.symbol);
         }
     }
+
+    private void createMapFunctionForStepExpr(BLangXMLNavigationAccess xmlNavigation,
+                                              List<BLangXMLStepExtend> extensions) {
+        Location pos = xmlNavigation.pos;
+        ArrayList<BLangExpression> args = new ArrayList<>();
+        ArrayList<BLangExpression> filters = expandFilters(xmlNavigation.filters);
+
+        // xml/**/<elemName>
+        if (xmlNavigation.navAccessType == XMLNavigationAccess.NavAccessType.DESCENDANTS) {
+            args.add(createArrowFunctionForNavigation(pos, extensions, XML_GET_DESCENDANTS, filters));
+        } else {
+            // xml/*
+            args.add(createArrowFunctionForNavigation(pos, extensions, XML_INTERNAL_CHILDREN, filters));
+        }
+
+        BLangInvocation elements =
+                createLanglibXMLInvocation(pos, XML_ELEMENTS, xmlNavigation.expr, new ArrayList<>(),
```

Comment:
Since `createLanglibXMLInvocation` doesn't seems to mutate the list given to it shall we change its parameters to accept `List<BLangExpression>` instead of `ArrayList`. Then lets use immutable lists (using `List.of` and `Collections.emptyList`) in the places where we call it instead of creating ArrayLists.